src/visit.rs

```diff
@@ -0,0 +1,236 @@
+//!         Item::Value(value) => v.visit_value(value),
+//!         Item::Table(table) => v.visit_table(table),
+//!         Item::ArrayOfTables(array) => v.visit_array_of_tables(array),
+//!     }
+//! }
+//! ```
+//!
+//! The API is modeled after [`syn::visit`](https://docs.rs/syn/1/syn/visit).
+//!
+//! # Examples
+//!
+//! This visitor stores every string in the document.
+//!
+//! ```
+//! # use toml_edit::*;
+//! use toml_edit::visit::*;
+//!
+//! #[derive(Default)]
+//! struct StringCollector<'doc> {
+//!     strings: Vec<&'doc str>,
+//! }
+//!
+//! impl<'doc> Visit<'doc> for StringCollector<'doc> {
+//!     fn visit_string(&mut self, node: &'doc Formatted<String>) {
+//!          self.strings.push(node.value().as_str());
+//!     }
+//! }
+//!
+//! let input = r#"
+//! laputa = "sky-castle"
+//! the-force = { value = "surrounds-you" }
+//! "#;
+//!
+//! let mut document: Document = input.parse().unwrap();
+//! let mut visitor = StringCollector::default();
+//! visitor.visit_document(&document);
+//!
+//! assert_eq!(visitor.strings, vec!["sky-castle", "surrounds-you"]);
+//! ```
+//!
+//! For a more complex example where the visitor has internal state, see `examples/visit.rs`
+//! [on GitHub](https://github.com/ordian/toml_edit/blob/master/examples/visit.rs).
+
+use crate::{
+    Array, ArrayOfTables, Datetime, Document, Formatted, InlineTable, Item, Table, TableLike, Value,
+};
+
+/// Document tree traversal to mutate an exclusive borrow of a document tree in-place.
+///
+/// See the [module documentation](self) for details.
+pub trait Visit<'doc> {
+    fn visit_array(&mut self, node: &'doc Array) {
+        visit_array(self, node);
+    }
+
+    fn visit_array_of_tables(&mut self, node: &'doc ArrayOfTables) {
+        visit_array_of_tables(self, node);
+    }
+
+    fn visit_boolean(&mut self, node: &'doc Formatted<bool>) {
+        visit_boolean(self, node)
+    }
+
+    fn visit_datetime(&mut self, node: &'doc Formatted<Datetime>) {
+        visit_datetime(self, node);
+    }
+
+    fn visit_document(&mut self, node: &'doc Document) {
+        visit_document(self, node);
+    }
+
+    fn visit_float(&mut self, node: &'doc Formatted<f64>) {
+        visit_float(self, node)
+    }
+
+    fn visit_inline_table(&mut self, node: &'doc InlineTable) {
+        visit_inline_table(self, node)
+    }
+
+    fn visit_integer(&mut self, node: &'doc Formatted<i64>) {
+        visit_integer(self, node)
+    }
+
+    fn visit_item(&mut self, node: &'doc Item) {
+        visit_item(self, node);
+    }
+
+    fn visit_string(&mut self, node: &'doc Formatted<String>) {
+        visit_string(self, node)
+    }
+
+    fn visit_table(&mut self, node: &'doc Table) {
+        visit_table(self, node);
+    }
+
+    fn visit_table_like(&mut self, node: &'doc dyn TableLike) {
+        visit_table_like(self, node);
+    }
+
+    fn visit_table_like_kv(&mut self, key: &'doc str, node: &'doc Item) {
+        visit_table_like_kv(self, key, node);
+    }
```

Comment:
It seems inconsistent that we expose `visit_table` but not `visit_inline_table`.   Was there a reason for this?

Reply:
Hmm I might be misunderstanding what you mean, but we do have a visit_inline_table method on the trait (the methods are in alphabetical order).

Reply:
You're right, I had missed it because I had assumed it would be grouped with them.  Maybe we should do that?  We grouped the array free and trait functions but all the tables are not grouped.

Reply:
I think that mostly happened because it fell out of the alphabetical order. But yes I can group the table functions together to minimize confusion, as well as the value functions.

Reply:
Addressed in the latest update.